src/Composer/DependencyResolver/PoolOptimizer.php

```diff
@@ -0,0 +1,279 @@
+                $this->addIrremovablePackageConstraint($package->getAliasOf()->getName(), new Constraint('==', $package->getAliasOf()->getVersion()));
+            }
+
+            // This is the lazy approach. We simply do not remove any packages that are replaced or replace others here.
+            // (maybe this can be improved?)
+            if ($package->getReplaces()) {
+                $this->addIrremovablePackageConstraint($package->getName(), new Constraint('==', $package->getVersion()));
+
+                foreach ($package->getReplaces() as $link) {
+                    // Make sure we do not replace ourselves (if someone made a mistake and tagged it)
+                    // See e.g. https://github.com/BabDev/Pagerfanta/commit/fd00eb74632fecc0265327e9fe0eddc08c72b238#diff-b5d0ee8c97c7abd7e3fa29b9a27d1780
+                    // TODO: should that go into package itself?
+                    if ($package->getName() === $link->getTarget()) {
+                        continue;
+                    }
+
+                    $this->addIrremovablePackageConstraint($link->getTarget(), $link->getConstraint());
+                }
+            }
+        }
+
+        // Mark the packages as irremovable based on the constraints
+        foreach ($this->pool->getPackages() as $i => $package) {
+            if (!isset($this->irremovablePackageConstraints[$package->getName()])) {
+                continue;
+            }
+
+            if (CompilingMatcher::match($this->irremovablePackageConstraints[$package->getName()], Constraint::OP_EQ, $package->getVersion())) {
+                $this->irremovablePackages[$package->id] = true;
+            }
+        }
+    }
+
+    private function addIrremovablePackageConstraint($packageName, ConstraintInterface $constraint)
+    {
+        if (!isset($this->irremovablePackageConstraints[$packageName])) {
+            $this->irremovablePackageConstraints[$packageName] = $constraint;
+
+            return;
+        }
+
+        // Do not use Intervals::compactConstraint() here (it hurts performance)
+        $this->irremovablePackageConstraints[$packageName] = new MultiConstraint(array(
+            $this->irremovablePackageConstraints[$packageName],
+            $constraint
+        ), false);
+    }
+
+    private function applyRemovalsToPool()
+    {
+        $packages = array();
+        foreach ($this->pool->getPackages() as $package) {
+            if (!isset($this->packagesToRemove[$package->id])) {
+                $packages[] = $package;
+            }
+        }
+
+        $this->pool = new Pool($packages, $this->pool->getUnacceptableFixedOrLockedPackages());
+
+        // Reset package removals
+        $this->packagesToRemove = array();
+    }
+
+    private function optimizeByIdenticalDependencies()
+    {
+        $identicalDefinitionPerPackage = array();
+        $packageIdsToRemove = array();
+
+        foreach ($this->pool->getPackages() as $i => $package) {
+
```

Comment:
If I get this correctly, it'd be IMO clearer and faster to do the check for irremovability here, so you can skip processing that package entirely.

```suggestion
        if (isset($this->irremovablePackages[$package->id])) {
            continue;
        }
```

Reply:
That's correct. Added in 2b004dbc3f10a05067b8a544c6baf9653c51965f.